File: tenant_stats.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
PostgreSQL 租户统计工具
功能：提取 SQL 中的 tenant_code，分析执行计划，计算预估流量并异步记录
"""

import re
import hashlib
import json
import asyncio
from datetime import datetime
from typing import Optional, Dict, Any, Tuple
import psycopg2
from psycopg2.extensions import ISOLATION_LEVEL_AUTOCOMMIT
import aiofiles


class TenantStatsCollector:
    """PostgreSQL 租户统计收集器"""
    
    def __init__(self, db_config: Dict[str, str], log_file: str = "tenant_stats.log"):
        """
        初始化收集器
        
        Args:
            db_config: 数据库连接配置 {host, port, database, user, password}
            log_file: 日志文件路径
        """
        self.db_config = db_config
        self.log_file = log_file
        self.cache: Dict[str, Dict[str, Any]] = {}  # MD5 -> explain result cache
        
    def extract_tenant_code(self, sql: str) -> Optional[str]:
        """
        从 SQL 中提取 tenant_code
        
        支持多种模式：
        - WHERE tenant_code = 'xxx'
        - WHERE tenant_code IN ('xxx')
        - AND tenant_code = 'xxx'
        - tenant_code::text = 'xxx'
        
        Args:
            sql: SQL 语句
            
        Returns:
            tenant_code 值，如果未找到则返回 None
        """
        # 移除注释和换行，便于匹配
        cleaned_sql = re.sub(r'--.*?$', '', sql, flags=re.MULTILINE)
        cleaned_sql = re.sub(r'/\*.*?\*/', '', cleaned_sql, flags=re.DOTALL)
        
        # 多种模式匹配
        patterns = [
            r"tenant_code\s*=\s*'([^']+)'",
            r"tenant_code\s*=\s*\"([^\"]+)\"",
            r"tenant_code\s+IN\s*\(\s*'([^']+)'",
            r"tenant_code::text\s*=\s*'([^']+)'",
        ]
        
        for pattern in patterns:
            match = re.search(pattern, cleaned_sql, re.IGNORECASE)
            if match:
                return match.group(1)
        
        return None
    
    def calculate_sql_md5(self, sql: str) -> str:
        """
        计算 SQL 的 MD5 哈希值
        
        Args:
            sql: SQL 语句
            
        Returns:
            MD5 哈希值（32位16进制字符串）
        """
        # 标准化 SQL：移除多余空格、换行，转小写
        normalized_sql = re.sub(r'\s+', ' ', sql.strip()).lower()
        return hashlib.md5(normalized_sql.encode('utf-8')).hexdigest()
    
    def get_explain_result(self, sql: str) -> Dict[str, Any]:
        """
        获取 SQL 的执行计划（支持 MD5 缓存）
        
        Args:
            sql: SQL 语句
            
        Returns:
            包含执行计划信息的字典
            {
                'total_cost': float,
                'plan_rows': int,
                'plan_width': int,
                'estimated_traffic': int,  # bytes
                'from_cache': bool
            }
        """
        sql_md5 = self.calculate_sql_md5(sql)
        
        # 检查缓存
        if sql_md5 in self.cache:
            result = self.cache[sql_md5].copy()
            result['from_cache'] = True
            return result
        
        # 执行 EXPLAIN
        conn = None
        try:
            conn = psycopg2.connect(**self.db_config)
            conn.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
            cursor = conn.cursor()
            
            # 执行 EXPLAIN (FORMAT JSON)
            explain_sql = f"EXPLAIN (FORMAT JSON) {sql}"
            cursor.execute(explain_sql)
            explain_result = cursor.fetchone()[0]
            
            # 解析结果
            plan = explain_result[0]['Plan']
            total_cost = plan['Total Cost']
            plan_rows = plan['Plan Rows']
            plan_width = plan['Plan Width']
            estimated_traffic = plan_rows * plan_width
            
            result = {
                'total_cost': total_cost,
                'plan_rows': plan_rows,
                'plan_width': plan_width,
                'estimated_traffic': estimated_traffic,
                'from_cache': False
            }
            
            # 存入缓存
            self.cache[sql_md5] = {
                'total_cost': total_cost,
                'plan_rows': plan_rows,
                'plan_width': plan_width,
                'estimated_traffic': estimated_traffic
            }
            
            return result
            
        except Exception as e:
            raise Exception(f"EXPLAIN 执行失败: {str(e)}")
        finally:
            if conn:
                conn.close()
    
    async def write_stats_async(self, tenant_code: str, stats: Dict[str, Any]):
        """
        异步写入统计结果到日志文件
        
        Args:
            tenant_code: 租户代码
            stats: 统计信息字典
        """
        timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        
        log_entry = {
            'timestamp': timestamp,
            'tenant_code': tenant_code,
            'total_cost': stats['total_cost'],
            'plan_rows': stats['plan_rows'],
            'plan_width': stats['plan_width'],
            'estimated_traffic_bytes': stats['estimated_traffic'],
            'estimated_traffic_mb': round(stats['estimated_traffic'] / 1024 / 1024, 2),
            'from_cache': stats['from_cache']
        }
        
        async with aiofiles.open(self.log_file, mode='a', encoding='utf-8') as f:
            await f.write(json.dumps(log_entry, ensure_ascii=False) + '\n')
    
    async def analyze_sql(self, sql: str) -> Dict[str, Any]:
        """
        完整分析流程：提取租户、获取执行计划、写入日志
        
        Args:
            sql: SQL 语句
            
        Returns:
            分析结果字典
        """
        # 1. 提取 tenant_code
        tenant_code = self.extract_tenant_code(sql)
        if not tenant_code:
            raise ValueError("无法从 SQL 中提取 tenant_code")
        
        # 2. 获取执行计划
        stats = self.get_explain_result(sql)
        
        # 3. 异步写入日志
        await self.write_stats_async(tenant_code, stats)
        
        # 4. 返回结果
        return {
            'tenant_code': tenant_code,
            **stats
        }
    
    def clear_cache(self):
        """清空 MD5 缓存"""
        self.cache.clear()
    
    def get_cache_size(self) -> int:
        """获取当前缓存大小"""
        return len(self.cache)


async def main():
    """示例使用"""
    
    # 数据库配置
    db_config = {
        'host': 'localhost',
        'port': '5432',
        'database': 'testdb',
        'user': 'postgres',
        'password': 'postgres'
    }
    
    # 创建收集器
    collector = TenantStatsCollector(db_config, log_file='tenant_stats.log')
    
    # 示例 SQL（包含多层 CTE）
    test_sql = """
    WITH tenant_orders AS (
        SELECT 
            order_id,
            user_id,
            amount,
            created_at
        FROM orders
        WHERE tenant_code = 'TENANT_001'
          AND status = 'completed'
    ),
    user_stats AS (
        SELECT 
            user_id,
            COUNT(*) as order_count,
            SUM(amount) as total_amount
        FROM tenant_orders
        GROUP BY user_id
    ),
    ranked_users AS (
        SELECT 
            user_id,
            order_count,
            total_amount,
            ROW_NUMBER() OVER (ORDER BY total_amount DESC) as rank
        FROM user_stats
    )
    SELECT 
        u.user_id,
        u.username,
        r.order_count,
        r.total_amount,
        r.rank
    FROM ranked_users r
    JOIN users u ON r.user_id = u.id
    WHERE r.rank <= 100
    ORDER BY r.rank;
    """
    
    try:
        # 第一次分析（从数据库获取）
        print("=== 第一次分析 ===")
        result1 = await collector.analyze_sql(test_sql)
        print(f"租户: {result1['tenant_code']}")
        print(f"代价: {result1['total_cost']}")
        print(f"预估行数: {result1['plan_rows']}")
        print(f"行宽: {result1['plan_width']}")
        print(f"预估流量: {result1['estimated_traffic']} bytes ({result1['estimated_traffic']/1024/1024:.2f} MB)")
        print(f"来自缓存: {result1['from_cache']}")
        print(f"当前缓存大小: {collector.get_cache_size()}\n")
        
        # 第二次分析（从缓存获取）
        print("=== 第二次分析（相同 SQL）===")
        result2 = await collector.analyze_sql(test_sql)
        print(f"租户: {result2['tenant_code']}")
        print(f"代价: {result2['total_cost']}")
        print(f"预估流量: {result2['estimated_traffic']} bytes")
        print(f"来自缓存: {result2['from_cache']}")
        print(f"当前缓存大小: {collector.get_cache_size()}\n")
        
        print(f"✓ 统计结果已写入 {collector.log_file}")
        
    except Exception as e:
        print(f"✗ 错误: {str(e)}")


if __name__ == "__main__":
    asyncio.run(main())

File: tenant_stats.go
package main

import (
	"context"
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"os"
	"regexp"
	"strings"
	"sync"
	"time"

	"github.com/jackc/pgx/v5"
)

// ExplainResult 执行计划结果
type ExplainResult struct {
	TotalCost        float64 `json:"total_cost"`
	PlanRows         int64   `json:"plan_rows"`
	PlanWidth        int     `json:"plan_width"`
	EstimatedTraffic int64   `json:"estimated_traffic"`
	FromCache        bool    `json:"from_cache"`
}

// StatsLog 统计日志结构
type StatsLog struct {
	Timestamp            string  `json:"timestamp"`
	TenantCode           string  `json:"tenant_code"`
	TotalCost            float64 `json:"total_cost"`
	PlanRows             int64   `json:"plan_rows"`
	PlanWidth            int     `json:"plan_width"`
	EstimatedTrafficByte int64   `json:"estimated_traffic_bytes"`
	EstimatedTrafficMB   float64 `json:"estimated_traffic_mb"`
	FromCache            bool    `json:"from_cache"`
}

// TenantStatsCollector PostgreSQL 租户统计收集器
type TenantStatsCollector struct {
	connString string
	logFile    string
	cache      map[string]*ExplainResult
	cacheMu    sync.RWMutex
	logMu      sync.Mutex
}

// NewTenantStatsCollector 创建新的收集器
func NewTenantStatsCollector(connString, logFile string) *TenantStatsCollector {
	return &TenantStatsCollector{
		connString: connString,
		logFile:    logFile,
		cache:      make(map[string]*ExplainResult),
	}
}

// ExtractTenantCode 从 SQL 中提取 tenant_code
func (c *TenantStatsCollector) ExtractTenantCode(sql string) (string, error) {
	// 移除注释
	sql = regexp.MustCompile(`--.*?$`).ReplaceAllString(sql, "")
	sql = regexp.MustCompile(`(?s)/\*.*?\*/`).ReplaceAllString(sql, "")

	// 多种模式匹配
	patterns := []string{
		`tenant_code\s*=\s*'([^']+)'`,
		`tenant_code\s*=\s*"([^"]+)"`,
		`tenant_code\s+IN\s*\(\s*'([^']+)'`,
		`tenant_code::text\s*=\s*'([^']+)'`,
	}

	for _, pattern := range patterns {
		re := regexp.MustCompile(`(?i)` + pattern)
		if matches := re.FindStringSubmatch(sql); matches != nil {
			return matches[1], nil
		}
	}

	return "", fmt.Errorf("无法从 SQL 中提取 tenant_code")
}

// CalculateSQLMD5 计算 SQL 的 MD5 哈希值
func (c *TenantStatsCollector) CalculateSQLMD5(sql string) string {
	// 标准化 SQL
	normalized := strings.ToLower(strings.TrimSpace(sql))
	normalized = regexp.MustCompile(`\s+`).ReplaceAllString(normalized, " ")

	// 计算 MD5
	hash := md5.Sum([]byte(normalized))
	return hex.EncodeToString(hash[:])
}

// GetExplainResult 获取执行计划（带缓存）
func (c *TenantStatsCollector) GetExplainResult(ctx context.Context, sql string) (*ExplainResult, error) {
	sqlMD5 := c.CalculateSQLMD5(sql)

	// 检查缓存
	c.cacheMu.RLock()
	if cached, ok := c.cache[sqlMD5]; ok {
		c.cacheMu.RUnlock()
		result := *cached
		result.FromCache = true
		return &result, nil
	}
	c.cacheMu.RUnlock()

	// 执行 EXPLAIN
	conn, err := pgx.Connect(ctx, c.connString)
	if err != nil {
		return nil, fmt.Errorf("数据库连接失败: %w", err)
	}
	defer conn.Close(ctx)

	explainSQL := fmt.Sprintf("EXPLAIN (FORMAT JSON) %s", sql)
	var jsonResult string
	err = conn.QueryRow(ctx, explainSQL).Scan(&jsonResult)
	if err != nil {
		return nil, fmt.Errorf("EXPLAIN 执行失败: %w", err)
	}

	// 解析 JSON
	var explainData []map[string]interface{}
	if err := json.Unmarshal([]byte(jsonResult), &explainData); err != nil {
		return nil, fmt.Errorf("JSON 解析失败: %w", err)
	}

	plan := explainData[0]["Plan"].(map[string]interface{})
	totalCost := plan["Total Cost"].(float64)
	planRows := int64(plan["Plan Rows"].(float64))
	planWidth := int(plan["Plan Width"].(float64))
	estimatedTraffic := planRows * int64(planWidth)

	result := &ExplainResult{
		TotalCost:        totalCost,
		PlanRows:         planRows,
		PlanWidth:        planWidth,
		EstimatedTraffic: estimatedTraffic,
		FromCache:        false,
	}

	// 存入缓存
	c.cacheMu.Lock()
	c.cache[sqlMD5] = &ExplainResult{
		TotalCost:        totalCost,
		PlanRows:         planRows,
		PlanWidth:        planWidth,
		EstimatedTraffic: estimatedTraffic,
	}
	c.cacheMu.Unlock()

	return result, nil
}

// WriteStatsAsync 异步写入统计结果
func (c *TenantStatsCollector) WriteStatsAsync(tenantCode string, stats *ExplainResult) error {
	c.logMu.Lock()
	defer c.logMu.Unlock()

	logEntry := StatsLog{
		Timestamp:            time.Now().Format("2006-01-02 15:04:05"),
		TenantCode:           tenantCode,
		TotalCost:            stats.TotalCost,
		PlanRows:             stats.PlanRows,
		PlanWidth:            stats.PlanWidth,
		EstimatedTrafficByte: stats.EstimatedTraffic,
		EstimatedTrafficMB:   float64(stats.EstimatedTraffic) / 1024 / 1024,
		FromCache:            stats.FromCache,
	}

	jsonData, err := json.Marshal(logEntry)
	if err != nil {
		return fmt.Errorf("JSON 序列化失败: %w", err)
	}

	f, err := os.OpenFile(c.logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return fmt.Errorf("打开日志文件失败: %w", err)
	}
	defer f.Close()

	if _, err := f.Write(append(jsonData, '\n')); err != nil {
		return fmt.Errorf("写入日志失败: %w", err)
	}

	return nil
}

// AnalyzeSQL 完整分析流程
func (c *TenantStatsCollector) AnalyzeSQL(ctx context.Context, sql string) (map[string]interface{}, error) {
	// 1. 提取 tenant_code
	tenantCode, err := c.ExtractTenantCode(sql)
	if err != nil {
		return nil, err
	}

	// 2. 获取执行计划
	stats, err := c.GetExplainResult(ctx, sql)
	if err != nil {
		return nil, err
	}

	// 3. 异步写入日志
	go func() {
		if err := c.WriteStatsAsync(tenantCode, stats); err != nil {
			log.Printf("写入日志失败: %v", err)
		}
	}()

	// 4. 返回结果
	return map[string]interface{}{
		"tenant_code":        tenantCode,
		"total_cost":         stats.TotalCost,
		"plan_rows":          stats.PlanRows,
		"plan_width":         stats.PlanWidth,
		"estimated_traffic":  stats.EstimatedTraffic,
		"from_cache":         stats.FromCache,
	}, nil
}

// ClearCache 清空缓存
func (c *TenantStatsCollector) ClearCache() {
	c.cacheMu.Lock()
	defer c.cacheMu.Unlock()
	c.cache = make(map[string]*ExplainResult)
}

// GetCacheSize 获取缓存大小
func (c *TenantStatsCollector) GetCacheSize() int {
	c.cacheMu.RLock()
	defer c.cacheMu.RUnlock()
	return len(c.cache)
}

func main() {
	// 数据库连接字符串
	connString := "postgres://postgres:postgres@localhost:5432/testdb"

	// 创建收集器
	collector := NewTenantStatsCollector(connString, "tenant_stats.log")

	// 示例 SQL
	testSQL := `
	WITH tenant_orders AS (
		SELECT 
			order_id,
			user_id,
			amount,
			created_at
		FROM orders
		WHERE tenant_code = 'TENANT_001'
		  AND status = 'completed'
	),
	user_stats AS (
		SELECT 
			user_id,
			COUNT(*) as order_count,
			SUM(amount) as total_amount
		FROM tenant_orders
		GROUP BY user_id
	)
	SELECT * FROM user_stats;
	`

	ctx := context.Background()

	// 第一次分析
	fmt.Println("=== 第一次分析 ===")
	result1, err := collector.AnalyzeSQL(ctx, testSQL)
	if err != nil {
		log.Fatalf("分析失败: %v", err)
	}
	printResult(result1)
	fmt.Printf("当前缓存大小: %d\n\n", collector.GetCacheSize())

	// 第二次分析（从缓存获取）
	time.Sleep(100 * time.Millisecond) // 等待异步写入完成
	fmt.Println("=== 第二次分析（相同 SQL）===")
	result2, err := collector.AnalyzeSQL(ctx, testSQL)
	if err != nil {
		log.Fatalf("分析失败: %v", err)
	}
	printResult(result2)
	fmt.Printf("当前缓存大小: %d\n", collector.GetCacheSize())

	fmt.Printf("\n✓ 统计结果已写入 %s\n", collector.logFile)
}

func printResult(result map[string]interface{}) {
	fmt.Printf("租户: %s\n", result["tenant_code"])
	fmt.Printf("代价: %.2f\n", result["total_cost"])
	fmt.Printf("预估行数: %d\n", result["plan_rows"])
	fmt.Printf("行宽: %d\n", result["plan_width"])
	traffic := result["estimated_traffic"].(int64)
	fmt.Printf("预估流量: %d bytes (%.2f MB)\n", traffic, float64(traffic)/1024/1024)
	fmt.Printf("来自缓存: %v\n", result["from_cache"])
}
